// eslint/lib/rules/no-useless-computed-key-compiled.js
/**
 * @fileoverview Rule to disallow unnecessary computed property keys in object literals
 * @author Burak Yigit Kaya
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var MESSAGE_UNNECESSARY_COMPUTED = "Unnecessarily computed property [{{property}}] found.";

module.exports = {
    meta: {
        docs: {
            description: "disallow unnecessary computed property keys in object literals",
            category: "ECMAScript 6",
            recommended: false
        }
    },
    create: function create(context) {
        var sourceCode = context.getSourceCode();

        return {
            Property: function Property(node) {
                if (!node.computed) {
                    return;
                }

                var key = node.key,
                    nodeType = _typeof(key.value);

                if (key.type === "Literal" && (nodeType === "string" || nodeType === "number")) {
                    context.report(node, MESSAGE_UNNECESSARY_COMPUTED, { property: sourceCode.getText(key) });
                }
            }
        };
    }
};
