// es5-ext/object/compare-compiled.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var strCompare = require('../string/#/case-insensitive-compare'),
    isObject = require('./is-object'),
    resolve,
    typeMap;

typeMap = {
	undefined: 0,
	object: 1,
	boolean: 2,
	string: 3,
	number: 4
};

resolve = function resolve(a) {
	if (isObject(a)) {
		if (typeof a.valueOf !== 'function') return NaN;
		a = a.valueOf();
		if (isObject(a)) {
			if (typeof a.toString !== 'function') return NaN;
			a = a.toString();
			if (typeof a !== 'string') return NaN;
		}
	}
	return a;
};

module.exports = function (a, b) {
	if (a === b) return 0; // Same

	a = resolve(a);
	b = resolve(b);
	if (a == b) return typeMap[typeof a === 'undefined' ? 'undefined' : _typeof(a)] - typeMap[typeof b === 'undefined' ? 'undefined' : _typeof(b)]; //jslint: ignore
	if (a == null) return -1;
	if (b == null) return 1;
	if (typeof a === 'string' || typeof b === 'string') {
		return strCompare.call(a, b);
	}
	if (a !== a && b !== b) return 0; //jslint: ignore
	return Number(a) - Number(b);
};
